resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -465,13 +532,40 @@ private boolean handleRedirect(
                         continue;
                     }
 
-                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
-                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                    final InetSocketAddress socketAddress = parent.newRedirectServerAddress(resolved);
+                    resolvedNameServers.add(socketAddress);
+
+                    addNameServerToCache(authoritativeNameServer, socketAddress, r.timeToLive());
                 }
+                final Set<InetSocketAddress> unresolvedNameServers;
+                if (serverNames.numUnhandled() > 0) {
+                    unresolvedNameServers = new LinkedHashSet<InetSocketAddress>(serverNames.numUnhandled());
+
+                    // Process all unresolved nameservers as well.
+                    for (;;) {
+                        final AuthoritativeNameServer authoritativeNameServer = serverNames.handleNext();
+                        if (authoritativeNameServer == null) {
```

Comment:
nit: you could use a do/while loop here because you already checked there is at least one element.